pkg/codegen/customresourcegenerator_test.go

```diff
@@ -0,0 +1,232 @@
+					Names: extensionv1.CustomResourceDefinitionNames{
+						Kind:     "TestKind",
+						Singular: "customsingular",
+						ListKind: "CustomListKind",
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			setCRDDefaults(&tt.crd)
+			if tt.crd.Spec.Names.Singular != tt.expected.Spec.Names.Singular {
+				t.Errorf("expected Singular %s, got %s", tt.expected.Spec.Names.Singular, tt.crd.Spec.Names.Singular)
+			}
+			if tt.crd.Spec.Names.ListKind != tt.expected.Spec.Names.ListKind {
+				t.Errorf("expected ListKind %s, got %s", tt.expected.Spec.Names.ListKind, tt.crd.Spec.Names.ListKind)
+			}
+		})
+	}
+}
+func TestNewCustomResourceGenerator(t *testing.T) {
+	tests := []struct {
+		name     string
+		crd      extensionv1.CustomResourceDefinition
+		expected CustomResourceGenerator
+		wantErr  bool
+	}{
+		{
+			name: "Valid CRD",
+			crd: extensionv1.CustomResourceDefinition{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "apiextensions.k8s.io/v1",
+				},
+				Spec: extensionv1.CustomResourceDefinitionSpec{
+					Group: "example.com",
+					Names: extensionv1.CustomResourceDefinitionNames{
+						Kind:   "TestKind",
+						Plural: "testkinds",
+					},
+					Versions: []extensionv1.CustomResourceDefinitionVersion{
+						{Name: "v1"},
+					},
+				},
+			},
+			expected: CustomResourceGenerator{
+				APIVersion: "apiextensions.k8s.io/v1",
+				Kind:       "TestKind",
+				Plural:     "testkinds",
+				Group:      "example.com",
+				Versions:   []string{"v1"},
+				GroupVersions: []string{
+					"example.com/v1",
+				},
+				ResourceTokens: []string{
+					"example.com:TestKind:v1",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "Valid CRD with multiple versions",
+			crd: extensionv1.CustomResourceDefinition{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: "apiextensions.k8s.io/v1",
+				},
+				Spec: extensionv1.CustomResourceDefinitionSpec{
+					Group: "example.com",
+					Names: extensionv1.CustomResourceDefinitionNames{
+						Kind:   "TestKind",
+						Plural: "testkinds",
+					},
+					Versions: []extensionv1.CustomResourceDefinitionVersion{
+						{Name: "v1"},
+						{Name: "v1alpha1"},
+					},
+				},
+			},
+			expected: CustomResourceGenerator{
+				APIVersion: "apiextensions.k8s.io/v1",
+				Kind:       "TestKind",
+				Plural:     "testkinds",
+				Group:      "example.com",
+				Versions:   []string{"v1alpha1", "v1"},
+				GroupVersions: []string{
+					"example.com/v1",
+					"example.com/v1alpha1",
+				},
+				ResourceTokens: []string{
+					"example.com:TestKind:v1",
+					"example.com:TestKind:v1alpha1",
+				},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := NewCustomResourceGenerator(tt.crd)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewCustomResourceGenerator() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !tt.wantErr {
```

Comment:
Nit: `if tt.wantErr != nil { continue }` to de-dent below, since you already checked it above.